missing-prototypes

NaruFGT opened the pull request:
Contributor @strager had recommended enabling missing-prototypes, which proved useful for identifying unused code. In addition to missing-prototypes in C code, this will enable -Wall and -Werror.

NaruFGT committed:
missing-prototypes
Contributor @strager had recommended enabling missing-prototypes, which proved us eful for identifying unused code. In addition to missing-prototypes, this will en able -Wall and -Werror for C code.

sas commented on CMakeLists.txt:
```diff
@@ -66,6 +66,8 @@ elseif ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang" OR "${CMAKE_CXX_COMPILER_ID}
     add_compile_options(-Wno-multichar -Wno-sign-compare)
   endif ()
+elseif ("${CMAKE_C_COMPILER_ID}" STREQUAL "GNU" OR "${CMAKE_C_COMPILER_ID}" STREQUAL "Clang")
+  add_compile_options(-Wall -Werror -Wmissing-prototypes)
```
I think this block will never be used. We already have an almost identical check at line 64. You wanna put `-Wmissing-prototype` at line 65.

NaruFGT replied:
The difference is whether we are invoking a C compiler or a C++ compiler. There are some libraries in this project which invoke the C compiler and CMake does not append C++ flags to invocations of the C compiler.
We cannot add -Wmissing-prototype to C++ CFLAGS because there are no "prototypes" in C++ and g++ will warn us which in our case with -Werror means those are errors.

NaruFGT replied:
While it does need to distinguish whether this is a C compiler or a C++ compiler, I see that  add_compile_options() may not help with this discernment, I'm reviewing CMake documentation, I'll check to see if this is the correct approach.

NaruFGT replied:
This is resolved with [this commit](https://github.com/facebook/xcbuild/pull/312/commits/32e37e9047c51deb25572ffd53331fab9bc83085).

NaruFGT committed:
Revert "missing-prototypes"
add_compile_options() will need CMake generator expressions to distinguish between invocations of C and C++ compilers. This commit did not solve the issue and may cause build errors if the missing-prototypes option is passed to C++ compilers.

This reverts commit 45a13570f1ad7178a232230510e4dcbafa9d71d8.

NaruFGT committed:
missing-prototypes & missing-declarations
Contributor @strager had recommended enabling missing-prototypes, which proved useful for identifying unused code.
I have added warning options -Wmissing-declarations and for the C compiler -Wmissing-prototypes.
Until these warnings are taken care of, I've disabled errors for these warnings.

sas reviewed (changes requested):
I don't think we should add these checks and then lower them to warning status unless we actually fix the issues in the code. The value provided by these warnings is low because it will just result in more noise during the build.

NaruFGT commented:
If we aren't going to address them then perhaps just disabling this warning entirely on GCC and Clang builds would be appropriate?

sas commented:
Disabling them completely would be fine by me. It's up to you; if you want we can disable them, but if you prefer fixing them and landing the fixes along with the changes that enable the warnings everywhere, then that works too.

Forgive me if I asked you this question before but are these warnings enabled by default on certain versions of gcc/clang? If I remember correctly, they are disabled by default and the user needs to enabled them.
